### PR TITLE
Add cancellation functionality to progress

### PR DESCRIPTION
**Pull request by gselzer**
# Description

This PR provides the ability for users to cancel tasks utilizing the `progress` mechanism. Therefore, if you write something like

```python
for i in progress(range(10)):
        do_something_time_consuming()
```
and the user decides to cancel on the fifth iteration of the `for` loop, the remaining 5 iterations will **not** be run. The user can assert cancellation with the field `progress.is_canceled()`, and can provide cancellation error-handling by passing a `Callable` to the `progress` object as the `cancel_callback` parameter - this can enable context destruction in the case of cancellation.

Open questions:
* What should the behavior be when there are inevitably lines beyond the `for` loop which expect the loop to have completed as expected? I see two possibilities:
  * We raise an exception of some sort within `progress.cancel`.
  * We require the remaining code to be conditional on `progress.is_cancel` returning `False

  I prefer the former, as it could be tough to check that if we don't retain a reference to progress. Retaining that reference would increase verbosity.
* Should we provide any kwargs to the cancellation callback? The cancellation callback might want to know e.g. how far computation got before it was cancelled. But I don't know what you'd want. We could also wait until someone requests that functionality to implement it.

# References

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How has this been tested?

- [X] test suite for my feature covers cases x, y, and z
- [X] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
I suggest defining something like `cancellable_progress` as many existing codes may end with a corrupted state that may require a restart of the whole application.

**Comment by gselzer** [comment]
> I suggest defining something like `cancellable_progress` as many existing codes may end with a corrupted state that may require a restart of the whole application.

Yeah, I had thought about that, but forgot to add it into my open questions section. Thanks for the catch

Would you prefer a new class (a subclass of `progress` would be minimal boilerplate, I think), or a kwarg for opt-in cancellation to minimize the number of classes? I don't feel strongly...

**Comment by Czaki** [comment]
I prefer new class

**Comment by gselzer** [comment]
@Czaki I took your suggestion and wrote a new class, but it feels clunky right now. Do you have any suggestions?

File: napari/_qt/widgets/qt_progress_bar.py [hunk 1]
```diff
@@ -44,6 +46,11 @@ def __init__(
 
         self.setLayout(base_layout)
 
+        if isinstance(prog, cancelable_progress):
+            self.cancel_button = QPushButton(trans._('Cancel'))
+            self.cancel_button.clicked.connect(self._cancel)
+            pbar_layout.addWidget(self.cancel_button)
+
```
Comment (Czaki) [inline napari/_qt/widgets/qt_progress_bar.py]: Conditional adding attribute is not the best practice. Maybe `self.cacnel_button.setVisible(isinstance(prog, cancelable_progress))`

File: napari/utils/progress.py [hunk 2]
```diff
@@ -150,3 +150,42 @@ def progrange(*args, **kwargs):
...
+        class Wrapped(Iterator):
+            def __iter__(wrapped):
+                return wrapped
+
+            def __next__(wrapped):
+                if self.is_canceled:
+                    raise StopIteration
+                return next(plain_iterator)
+
+        return Wrapped()
```
Comment (Czaki) [inline napari/utils/progress.py]: Maybe takewhile may be used? https://docs.python.org/3/library/itertools.html#itertools.takewhile

**Review by Czaki — COMMENTED** [review]
did I correctly found "clunky" places

File: napari/utils/progress.py (hunk 2 above)
Comment (gselzer) [reply napari/utils/progress.py]: Great idea, fixed with 6b4ddab

File: napari/_qt/widgets/qt_progress_bar.py (hunk 1 above)
Comment (gselzer) [reply napari/_qt/widgets/qt_progress_bar.py]: Added it in cb34bc93

**Comment by gselzer** [comment]
@Czaki I fixed the points you suggested ~~(although for some reason my [latest commit](https://github.com/napari/napari/compare/main...gselzer:napari:progress_cancellation) isn't showing on this PR)~~, let me know if there's any other blockers to merge? @DragaDoncila did you get a chance to look at this?

EDIT: The commit showed up, it just took like 20 minutes

File: napari/utils/progress.py [hunk 3]
```diff
@@ -150,3 +151,30 @@ def progrange(*args, **kwargs):
...
+        *args,
+        **kwargs,
+    ) -> None:
+        self.cancel_callback = cancel_callback
+        self.is_canceled = False
+
+        super().__init__(iterable, desc, total, nest_under, *args, **kwargs)
+
+    def __iter__(self) -> Iterator:
+        return takewhile(lambda _: not self.is_canceled, self.iterable)
```
Comment (DragaDoncila) [inline napari/utils/progress.py]: This leaves us unable to see progress bar updates in the GUI, and the bar doesn't disappear after cancellation. I think it should be:

```suggestion
        return takewhile(lambda _: not self.is_canceled, super().__iter__())
```

**Review by DragaDoncila — CHANGES_REQUESTED** [review]
@gselzer thanks for opening this! I've had a chance to have a look now and I like the implementation. I agree with Czaki's original suggestion that it should be its own class.

I left a comment on the `takewhile` line - we need to pass it the super-class' iterator so that it can emit the correct events for updating the GUI - currently the GUI progress bar doesn't do anything.

It would be great to add a `cancelable_progress` example to the `progress_minimal.py` example e.g.

```python
def cancel_callback():
    print("Canceling") # would be nice if this was something more useful but.. not sure what it could be
    
def cancelable_iterable():
    for _ in cancelable_progress(range(100), cancel_callback=cancel_callback):
        np.random.rand(128, 128, 128).mean(0)
        
cancel_iter_btn = QPushButton("Cancelable Iterable")
cancel_iter_btn.clicked.connect(cancelable_iterable)
button_layout.addWidget(cancel_iter_btn)
```

Otherwise looking good! Next step would then be to add the cancel functionality to the generator threadworkers in `_qt/threading/create_worker`, but that's a follow-up PR imo.

**Comment by gselzer** [comment]
Thanks @DragaDoncila!

I made your requested changes.

I had one more concern (beyond the failing CI checks, which I can't figure out how to fix):

"Could the user cancel *during* an iteration of the for loop, causing the cancel_callback to be triggered in the middle of an iteration?" Could this cause errors?

I wonder if we could instead call `cancel_callback` within the `takewhile`, so that `cancel_callback` doesn't ever run in the middle of an execution? Something like

```python
def __iter__(self):
  def check_canceled():
    if self.is_canceled:
      self.cancel_callback()
      return False
    return True
  return takewhile(check_canceled, super().__iter__())
```

EDIT: I made this change in 0e438aa. Let me know what you think

**Comment by gselzer** [comment]
@DragaDoncila is there anything else I can do to get this merged?

**Review by DragaDoncila — APPROVED** [review]
Thank you for making those changes @gselzer! Agreed, checking for cancelation in the `takewhile` makes more sense and is less likely to lead to chaos. This looks good to me, excited to get it in! @Czaki did you want to take a final pass?

**Comment by gselzer** [comment]
@DragaDoncila @Czaki anything else I can do to get this merged?

File: napari/utils/progress.py [hunk 4]
```diff
@@ -150,3 +151,59 @@ def progrange(*args, **kwargs):
...
+                # If we've canceled, run the callback and then notify takewhile
+                if self.cancel_callback:
+                    self.cancel_callback()
+                return False
+                # Otherwise, continue
+            return True
+
+        return takewhile(is_canceled, super().__iter__())
+
+    def cancel(self):
```
Comment (Czaki) [inline napari/utils/progress.py:213]: This method should have a docstring to be visible in API documentation.

**Comment by Czaki** [comment]
The general problem with the current mechanism is that if the generator acquire some resources then I do not know when it will be released. 

I read a few things about generators, and it looks like the [`generator.close()`](https://docs.python.org/3/reference/expressions.html#generator.close) should be used. 

But I do not have time to test if I'm correct and how important is this problem

File: napari/utils/progress.py (hunk 4 above)
Comment (gselzer) [reply napari/utils/progress.py:213]: Docstring added in fb518dc8

**Comment by gselzer** [comment]
@Czaki so I wrote [558aa20](https://github.com/napari/napari/pull/5728/commits/558aa20fc4d9d36fac66001125d02c06cbc4eca2) to try and address your concern about closing a generator, but I can't seem to figure out how you close the passed generator. The test that I wrote fails because calling `close()` on the iterable doesn't actually make its way to the `Generator` wrapped by `tqdm`. Is this an oversight in `tqdm`?

I personally don't know enough about these situations in Python to think of an example where the cleanup shouldn't be done in the `cancel_callback` parameter. Do you have an example in mind?

File: napari/utils/_tests/test_progress.py [hunk 5]
```diff
@@ -93,3 +93,59 @@ def test_progrange():
...
+    closed = False
+
+    def test_generator():
+        try:
+            i = 0
+            while True:
+                yield i
+                i += 1
+        finally:
+            global closed
```
Comment (Czaki) [inline napari/utils/_tests/test_progress.py]:
```suggestion
            nonlocal closed
```

Closed is not a module-level variable.

**Review by Czaki — COMMENTED** [review]
Ok. There is a bug in tqdm. I will create PR.

File: napari/utils/progress.py [hunk 6]
```diff
@@ -150,3 +151,68 @@ def progrange(*args, **kwargs):
...
+        itr = super().__iter__()
+
+        def is_canceled(_):
+            if self.is_canceled:
+                # If we've canceled, run the callback and then notify takewhile
+                if self.cancel_callback:
+                    self.cancel_callback()
+                    # Perform additional cleanup for generators
+                    if isinstance(itr, Generator):
+                        itr.close()
```
Comment (Czaki) [inline napari/utils/progress.py]:
```suggestion
                    if isinstance(self.iterable, Generator):
                        self.iterable.close()
```

The `tqdm` exposes the inner iterable object on attribute `iterable`.

**Review by Czaki — COMMENTED** [review]
After local testing, it looks like these two changes should be enough to get the code working.

**Comment by gselzer** [comment]
@Czaki I think I fixed everything you mentioned, let me know if there's anything else!

**Review by Czaki — APPROVED** [review]

**Comment by GenevieveBuckley** [comment]
This PR has two approvals, all the CI checks are green, and is marked ready to merge. Merging!

Thank you so much for your work on this @gselzer

**Pull request merged**